File: hw4/src/ru/hse/OperationResponse.py
import base64
import pickle
from typing import Any, Optional, Union

import javaobj
from javaobj import load
from io import BytesIO

from javaobj.v2.transformers import JavaInt


class OperationResponse:
    # Константы кодов
    SUCCEED = 0
    ALREADY_LOGGED = 1
    NOT_LOGGED = 2
    NO_USER_INCORRECT_PASSWORD = 3
    INCORRECT_RESPONSE = 4
    UNDEFINED_ERROR = 5
    INCORRECT_SESSION = 6
    NO_MONEY = 7
    ENCODING_ERROR = 8
    ALREADY_INITIATED = 9
    NULL_ARGUMENT = 10
    CONNECTION_ERROR = 11

    # Предопределенные экземпляры
    ACCOUNT_MANAGER_RESPONSE = None
    NO_USER_INCORRECT_PASSWORD_RESPONSE = None
    UNDEFINED_ERROR_RESPONSE = None
    NOT_LOGGED_RESPONSE = None
    INCORRECT_SESSION_RESPONSE = None
    SUCCEED_RESPONSE = None
    NO_MONEY_RESPONSE = None
    ENCODING_ERROR_RESPONSE = None
    ALREADY_INITIATED_RESPONSE = None
    NULL_ARGUMENT_EXCEPTION = None
    CONNECTION_ERROR_RESPONSE = None

    def __init__(self, code: int, body: Any = None):
        self.code = code
        self.body = body

    def to_result_string(self) -> str:
        sb = str(self.code)
        if self.body is not None:
            if isinstance(self.body, OperationResponse):
                sb += "|3|" + self.body.to_result_string()
            elif isinstance(self.body, str):
                sb += "|1|" + self.body
            else:
                sb += "|2|" + self._response_to_string(self.body)
        return sb

    @classmethod
    def from_string(cls, string: str) -> 'OperationResponse':
        ind = string.find("|")
        try:
            if ind > -1:
                code = int(string[:ind])
                string = string[ind+1:]
                ind = string.find("|")
                mode = int(string[:ind])
                if mode == 1:
                    return OperationResponse(code, string[ind+1:])
                elif mode == 2:
                    return OperationResponse(code, cls._response_from_string(string[ind+1:]))
                elif mode == 3:
                    return OperationResponse(code, cls.from_string(string[ind+1:]))

            code = int(string)
            if code >= 0:
                return OperationResponse(code, None)
        except (ValueError, Exception):
            pass

        return OperationResponse(cls.UNDEFINED_ERROR, string)

    @classmethod
    def code_to_error_message(cls, code: int) -> str:
        messages = {
            cls.SUCCEED: "SUCCEED",
            cls.ALREADY_LOGGED: "ALREADY LOGGED OR REGISTERED",
            cls.NOT_LOGGED: "NOT LOGGED",
            cls.NO_USER_INCORRECT_PASSWORD: "INCORRECT PASSWORD OR NO SUCH USER",
            cls.INCORRECT_RESPONSE: "INCORRECT RESPONCE",
            cls.UNDEFINED_ERROR: "UNDEFINED ERROR",
            cls.INCORRECT_SESSION: "INCORRECT SESSION NUMBER",
            cls.NO_MONEY: "NOT ENOUGH MONEYS ON BALANCE",
            cls.ENCODING_ERROR: "ENCODING CANNOT BE MADE",
            cls.ALREADY_INITIATED: "ACCOUNT WAS ALREADY INITIATED",
            cls.NULL_ARGUMENT: "Null argument is prohibited",
            cls.CONNECTION_ERROR: "No connection to server"
        }
        return messages.get(code, f"CODE_{code}")

    def __str__(self) -> str:
        sb = self.code_to_error_message(self.code)
        if self.body is not None:
            sb += f"[{str(self.body)}]"
        return sb

    @staticmethod
    def _response_to_string(obj: Any) -> str:
        try:
            serialized = pickle.dumps(obj)
            return base64.b64encode(serialized).decode('utf-8')
        except Exception as e:
            raise RuntimeError(f"Serialization error: {e}")

    @staticmethod
    def _response_from_string(string: str) -> Any:
        try:
            data = base64.b64decode(string)
            file_like = BytesIO(data)
            java_obj = load(file_like)
            return java_obj.value
        except Exception as e:
            raise RuntimeError(f"Deserialization error: {e}")

# Инициализация статических экземпляров после определения класса
OperationResponse.ACCOUNT_MANAGER_RESPONSE = OperationResponse(OperationResponse.ALREADY_LOGGED, None)
OperationResponse.NO_USER_INCORRECT_PASSWORD_RESPONSE = OperationResponse(OperationResponse.NO_USER_INCORRECT_PASSWORD, None)
OperationResponse.UNDEFINED_ERROR_RESPONSE = OperationResponse(OperationResponse.UNDEFINED_ERROR, None)
OperationResponse.NOT_LOGGED_RESPONSE = OperationResponse(OperationResponse.NOT_LOGGED, None)
OperationResponse.INCORRECT_SESSION_RESPONSE = OperationResponse(OperationResponse.INCORRECT_SESSION, None)
OperationResponse.SUCCEED_RESPONSE = OperationResponse(OperationResponse.SUCCEED, None)
OperationResponse.NO_MONEY_RESPONSE = OperationResponse(OperationResponse.NO_MONEY, None)
OperationResponse.ENCODING_ERROR_RESPONSE = OperationResponse(OperationResponse.ENCODING_ERROR, None)
OperationResponse.ALREADY_INITIATED_RESPONSE = OperationResponse(OperationResponse.ALREADY_INITIATED, None)
OperationResponse.NULL_ARGUMENT_EXCEPTION = OperationResponse(OperationResponse.NULL_ARGUMENT, None)
OperationResponse.CONNECTION_ERROR_RESPONSE = OperationResponse(OperationResponse.CONNECTION_ERROR, None)

File: hw5/src/main/ru/hse/OperationResponse.java
package ru.hse;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.Base64;

public class OperationResponse implements Serializable {
    private static final long serialVersionUID = 1L;
    private static final char SEPARATOR = '|';
    private static final int STRING_BODY_MODE = 1;
    private static final int SERIALIZABLE_BODY_MODE = 2;
    private static final int RESPONSE_BODY_MODE = 3;

    public static final int SUCCEED = 0;
    public static final int ALREADY_LOGGED = 1;
    public static final int NOT_LOGGED = 2;
    public static final int NO_USER_INCORRECT_PASSWORD = 3;
    public static final int INCORRECT_RESPONSE = 4;
    public static final int UNDEFINED_ERROR = 5;
    public static final int INCORRECT_SESSION = 6;
    public static final int NO_MONEY = 7;
    public static final int ENCODING_ERROR = 8;
    public static final int ALREADY_INITIATED = 9;
    public static final int NULL_ARGUMENT = 10;
    public static final int CONNECTION_ERROR = 11;

    public static final OperationResponse ACCOUNT_MANAGER_RESPONSE = new OperationResponse(ALREADY_LOGGED, null);
    public static final OperationResponse NO_USER_INCORRECT_PASSWORD_RESPONSE = new OperationResponse(NO_USER_INCORRECT_PASSWORD, null);
    public static final OperationResponse UNDEFINED_ERROR_RESPONSE = new OperationResponse(UNDEFINED_ERROR, null);
    public static final OperationResponse NOT_LOGGED_RESPONSE = new OperationResponse(NOT_LOGGED, null);
    public static final OperationResponse INCORRECT_SESSION_RESPONSE = new OperationResponse(INCORRECT_SESSION, null);
    public static final OperationResponse SUCCEED_RESPONSE = new OperationResponse(SUCCEED, null);
    public static final OperationResponse NO_MONEY_RESPONSE = new OperationResponse(NO_MONEY, null);
    public static final OperationResponse ENCODING_ERROR_RESPONSE = new OperationResponse(ENCODING_ERROR, null);
    public static final OperationResponse ALREADY_INITIATED_RESPONSE = new OperationResponse(ALREADY_INITIATED, null);
    public static final OperationResponse NULL_ARGUMENT_EXCEPTION = new OperationResponse(NULL_ARGUMENT, null);
    public static final OperationResponse CONNECTION_ERROR_RESPONSE = new OperationResponse(CONNECTION_ERROR, null);

    public final int code;
    public final Object body;

    public OperationResponse(int code, Object obj) {
        this.code = code;
        this.body = obj;
    }

    public String toResultString() {
        StringBuilder builder = new StringBuilder().append(code);
        if (body == null) {
            return builder.toString();
        }
        if (body instanceof OperationResponse responseBody) {
            return builder.append(SEPARATOR).append(RESPONSE_BODY_MODE).append(SEPARATOR).append(responseBody.toResultString()).toString();
        }
        if (body instanceof String textBody) {
            return builder.append(SEPARATOR).append(STRING_BODY_MODE).append(SEPARATOR).append(textBody).toString();
        }
        if (body instanceof Serializable serializableBody) {
            try {
                return builder.append(SEPARATOR).append(SERIALIZABLE_BODY_MODE).append(SEPARATOR).append(responseToString(serializableBody)).toString();
            } catch (IOException exception) {
                throw new IllegalStateException("Unable to serialize operation response body", exception);
            }
        }
        throw new IllegalStateException("Operation response body must be serializable");
    }

    public static OperationResponse fromString(String rawResponse) {
        try {
            return parseResponse(rawResponse);
        } catch (NumberFormatException | IOException exception) {
            return new OperationResponse(UNDEFINED_ERROR, rawResponse);
        } catch (ClassNotFoundException exception) {
            throw new IllegalStateException("Unable to deserialize operation response body", exception);
        }
    }

    private static OperationResponse parseResponse(String rawResponse) throws IOException, ClassNotFoundException {
        int codeSeparator = rawResponse.indexOf(SEPARATOR);
        if (codeSeparator < 0) {
            return parseCodeOnly(rawResponse);
        }

        int parsedCode = Integer.parseInt(rawResponse.substring(0, codeSeparator));
        String rawBody = rawResponse.substring(codeSeparator + 1);
        int modeSeparator = rawBody.indexOf(SEPARATOR);
        if (modeSeparator < 0) {
            return new OperationResponse(UNDEFINED_ERROR, rawResponse);
        }

        int mode = Integer.parseInt(rawBody.substring(0, modeSeparator));
        String payload = rawBody.substring(modeSeparator + 1);
        return switch (mode) {
            case STRING_BODY_MODE -> new OperationResponse(parsedCode, payload);
            case SERIALIZABLE_BODY_MODE -> new OperationResponse(parsedCode, responseFromString(payload));
            case RESPONSE_BODY_MODE -> new OperationResponse(parsedCode, fromString(payload));
            default -> new OperationResponse(UNDEFINED_ERROR, rawResponse);
        };
    }

    private static OperationResponse parseCodeOnly(String rawResponse) {
        int parsedCode = Integer.parseInt(rawResponse);
        return new OperationResponse(parsedCode, null);
    }

    public static String codeToErrorMessage(int code) {
        return switch (code) {
            case SUCCEED -> "SUCCEED";
            case ALREADY_LOGGED -> "ALREADY LOGGED OR REGISTERED";
            case NOT_LOGGED -> "NOT LOGGED";
            case NO_USER_INCORRECT_PASSWORD -> "INCORRECT PASSWORD OR NO SUCH USER";
            case INCORRECT_RESPONSE -> "INCORRECT RESPONSE";
            case UNDEFINED_ERROR -> "UNDEFINED ERROR";
            case INCORRECT_SESSION -> "INCORRECT SESSION NUMBER";
            case NO_MONEY -> "NOT ENOUGH MONEY ON BALANCE";
            case ENCODING_ERROR -> "ENCODING CANNOT BE MADE";
            case ALREADY_INITIATED -> "ACCOUNT WAS ALREADY INITIATED";
            case NULL_ARGUMENT -> "Null argument is prohibited";
            case CONNECTION_ERROR -> "No connection to server";
            default -> "CODE_" + code;
        };
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder(codeToErrorMessage(code));
        if (body != null) {
            builder.append('[').append(body).append(']');
        }
        return builder.toString();
    }

    public static String responseToString(Serializable value) throws IOException {
        try (ByteArrayOutputStream byteStream = new ByteArrayOutputStream(); ObjectOutputStream objectStream = new ObjectOutputStream(byteStream)) {
            objectStream.writeObject(value);
            objectStream.flush();
            return Base64.getEncoder().encodeToString(byteStream.toByteArray());
        }
    }

    /**
     * Deserializes a Base64 encoded string back to an object.
     */
    @SuppressFBWarnings("SECOBDES")
    public static Object responseFromString(String encodedValue) throws IOException, ClassNotFoundException {
        byte[] data = Base64.getDecoder().decode(encodedValue);
        try (ObjectInputStream objectStream = new ObjectInputStream(new ByteArrayInputStream(data))) {
            return objectStream.readObject();
        }
    }
}
